haystack_experimental/components/tools/tool_invoker.py

```diff
@@ -0,0 +1,204 @@
+
+    At initialization, the ToolInvoker component is provided with a list of available tools.
+    At runtime, the component processes a ChatMessage object containing tool calls and invokes the corresponding tools.
+    The results of the tool invocations are returned as a list of ChatMessage objects with tool role.
+
+    Usage example:
+    ```python
+    from haystack_experimental.dataclasses import ChatMessage, ToolCall, Tool
+    from haystack_experimental.components.tools import ToolInvoker
+
+    # Tool definition
+    def dummy_weather_function(city: str):
+        return f"The weather in {city} is 20 degrees."
+
+    parameters = {"type": "object",
+                "properties": {"city": {"type": "string"}},
+                "required": ["city"]}
+
+    tool = Tool(name="weather_tool",
+                description="A tool to get the weather",
+                function=dummy_weather_function,
+                parameters=parameters)
+
+    # Usually, the ChatMessage with tool_calls is generated by a Language Model
+    # Here, we create it manually for demonstration purposes
+    tool_call = ToolCall(
+        tool_name="weather_tool",
+        arguments={"city": "Berlin"}
+    )
+    message = ChatMessage.from_assistant(tool_calls=[tool_call])
+
+    # ToolInvoker initialization and run
+    invoker = ToolInvoker(tools=[tool])
+    result = invoker.run(message=message)
+
+    print(result)
+    ```
+
+    ```
+    >>  {
+    >>      'tool_messages': [
+    >>          ChatMessage(
+    >>              _role=<ChatRole.TOOL: 'tool'>,
+    >>              _content=[
+    >>                  ToolCallResult(
+    >>                      result='"The weather in Berlin is 20 degrees."',
+    >>                      origin=ToolCall(
+    >>                          tool_name='weather_tool',
+    >>                          arguments={'city': 'Berlin'},
+    >>                          id=None
+    >>                      )
+    >>                  )
+    >>              ],
+    >>              _meta={}
+    >>          )
+    >>      ]
+    >>  }
+    ```
+    """
+
+    def __init__(self, tools: List[Tool], raise_on_failure: bool = True):
+        """
+        Initialize the ToolInvoker component.
+
+        :param tools:
+            A list of tools that can be invoked.
+        :param raise_on_failure:
+            If True, the component will raise exceptions in case the tool is not found or invocation fails.
+            If False, the component will return a ChatMessage from the tool, wrapping the error message.
+        """
+        if not tools:
+            raise ValueError("ToolInvoker requires at least one tool to be provided.")
+
+        self.tools = tools
+        self._tools_with_names = {tool.name: tool for tool in tools}
+        self.raise_on_failure = raise_on_failure
+
+    @staticmethod
+    def _convert_tool_result_to_string(result: Any) -> str:
```

Comment:
Why don't we use `raise_on_failure` here?

Reply:
this was my initial idea.
Then I realized that even if some (good) results are not JSON serializable, their string representation can be OK for humans/language models.
Think for example of a datetime object...

Reply:
Now we default to `str` and allow customizability via a dedicated init parameter.

For consistency, in case of errors, with `raise_on_failure=False`, we return a `ChatMessage` containing an error.